written/weeks4-5.md

```diff
@@ -1,23 +1,87 @@
 ---
-author: Justin Perona
-title: ECS 154B Written Problems for Weeks 4-5, Winter 2019
+Author: Justin Perona
+Title: ECS 154B Written Problems for Weeks 4-5, Winter 2019
 ---
 
 # Written problems for weeks 4-5
 
 This problem set covers lectures 11 through 14.
+(The week numbers are slightly off, due to the fact that the midterm was in the middle of the week.)
 
 Solutions won't be posted on GitHub.
 If you'd like answers to the problems, either post on Piazza or go to Jason's or Justin's office hours.
 
 ## Lectures
 
-11. [](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-01-Lecture-11.pdf)
-12. [](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-04-Lecture-12.pdf)
-13. [](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-06-Lecture-13.pdf)
-14. [](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-08-Lecture-14.pdf)
+11. [More pipelining](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-01-Lecture-11.pdf)
+12. [Hazards](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-04-Lecture-12.pdf)
+13. [Control hazards](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-06-Lecture-13.pdf)
+14. [Pipeline correctness](https://github.com/jlpteaching/ECS154B/blob/master/lecture%20notes/02-08-Lecture-14.pdf)
 
 ## Topics covered
 
+* Pipelining
+    * Performance implications
+    * Datapath modifications
+* Hazards
+    * Data hazards
+    * Control hazards
+    * Branch prediction
+* Exceptions
+    * Reasoning
+    * Types of exceptions
+
 ## Problems
 
+### Pipelining
+
+1. Consider our traditional 5-stage pipeline. Assume that we are in the middle of executing a completely unrealistic program with no data dependencies, no branch or jump instructions, and no exceptions occurring. What is the CPI of the processor?
+2. Now consider if we had just started the program in the previous problem. What is the CPI of the processor through the first 5 cycles? What happens to this number as we increase the number of cycles executed up to 10? Up to infinity?
+3. How does a pipelined processor compare to a single cycle processor for the following terms? Do they go up, down, or stay the same?
+    * amount of power consumed
+    * amount of time consumed
+    * amount of energy consumed
+4. We don't usually see extremely large pipelines. (Intel processors peaked at 31 stages in the Prescott and Cedar Mill microarchitectures around 2006.) Part of the reason is due to hold time. Why does hold time influence pipeline length? (Hint: what changes when we split a 5-stage pipeline into, say, 100 stages?)
```

Comment:
I'm not sure "hold time" is the right word. I didn't use it lecture. Maybe you mean setup time? Or time to fill the pipeline?